### PR TITLE
Add possibility to format generated C++ code with clang-format.

### PR DESCRIPTION
**Pull request by lucteo**
If `clang-format` is found in the path, allow the CXXCodeWriter to use it to format the output C++ code.

### CODE REVIEW

File: Sources/CodeGen/CXX/CodeTransform.swift [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/// A transformation that can be run on code.
+public typealias CodeTransform = (String) -> String
+
+/// Returns an identity code transform.
+public func identity() -> CodeTransform {
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CodeTransform.swift:5]: I'm a bit uncomfortable having `identity` be introduced in the global namespace, as a public symbol. I think there's a better way to initialize the code writer (see other comment).

File: Sources/CodeGen/CXX/CXXCodeWriter.swift [hunk 2]
```diff
@@ -4,7 +4,12 @@ import ValModule
 public struct CXXCodeWriter {
 
   /// Initializes the current object.
-  public init() {}
+  public init(formatter: @escaping CodeTransform = identity()) {
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CXXCodeWriter.swift:7]:
```suggestion
  public init(formatter: CodeTransform? = nil) {
```

IIUC, that's what your `identity` function is representing: the absence of any transformation.

File: Sources/CodeGen/CXX/CodeTransform.swift [hunk 3]
```diff
@@ -0,0 +1,7 @@
+/// A transformation that can be run on code.
+public typealias CodeTransform = (String) -> String
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CodeTransform.swift:2]: This should be in the namespace of `CXXCodeWriter`

File: Sources/CodeGen/CXX/CXXCodeWriter.swift [hunk 4]
```diff
@@ -110,7 +115,7 @@ public struct CXXCodeWriter {
       target.writeNewline()
     }
 
-    return target.code
+    return formatter(target.code)
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CXXCodeWriter.swift:118]:
```suggestion
    return formatter?(target.code)
```

File: Sources/CodeGen/CXX/CXXCodeWriter.swift [hunk 5]
```diff
@@ -69,7 +74,7 @@ public struct CXXCodeWriter {
       }
     }
 
-    return target.code
+    return formatter(target.code)
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CXXCodeWriter.swift:77]:
```suggestion
    return formatter?(target.code)
```

File: Sources/CodeGen/CXX/CXXCodeWriter.swift [hunk 6]
```diff
@@ -4,7 +4,12 @@ import ValModule
 public struct CXXCodeWriter {
 
   /// Initializes the current object.
-  public init() {}
+  public init(formatter: @escaping CodeTransform = identity()) {
+    self.formatter = formatter
+  }
+
+  /// Object used to format output C++ code.
+  private let formatter: CodeTransform
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/CXXCodeWriter.swift:12]:
```suggestion
  private let formatter: CodeTransform?
```

File: Sources/CodeGen/CXX/ClangFormatTransformation.swift [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import Foundation
+
+/// Returns a C++ code formatter that invokes external `clang-format` to perform the job.
+public func clangFormatTransformation(_ executable: URL, style: String = "llvm") -> (
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/ClangFormatTransformation.swift:4]:
```suggestion
public func makeClangFormatter(_ executable: URL, style: String = "llvm") -> (
```

File: Sources/ValCommand/ValCommand.swift [hunk 8]
```diff
@@ -187,9 +187,15 @@ public struct ValCommand: ParsableCommand {
 
     // *** C++ Transpiling ***
 
+    let clangFormat = try? find("clang-format")
+    let codeFormatter: CodeTransform =
+      clangFormat != nil
+      ? clangFormatTransformation(URL(fileURLWithPath: clangFormat!))
+      : identity()
+
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:195]:
```suggestion
    let codeFormatter: CodeTransform? = (try? find("clang-format")).map({ clangFormatTransformation(URL(fileURLWithPath: $0!)) })
```

Can be improved if `clangFormatTransformation` does the URL construction from a path.

**Review by kyouko-taiga — CHANGES_REQUESTED** [review]
Great work. I'm very excited about this change!

I think there's a way to avoid introducing an `identity` function in the global namespace. See my inline comments.

File: Sources/CodeGen/CXX/CodeTransform.swift (hunk 3 above)
Comment (lucteo) [reply Sources/CodeGen/CXX/CodeTransform.swift:2]: @kyouko-taiga : If `CXXCodeWriter` disappears in the near future, where do you think it's better to put `CodeTransform`?

For me, keeping it here doesn't seem that bad (after removing `identity()` function)

File: Sources/CodeGen/CXX/ClangFormatter.swift [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import Foundation
+
+/// Returns a C++ code formatter that invokes external `clang-format` to perform the job.
+public func clangFormatter(_ executable: URL, style: String = "llvm") -> (
+  (String) -> String
+) {
+  return { code in
```
Comment (kyouko-taiga) [inline Sources/CodeGen/CXX/ClangFormatter.swift:7]:
```suggestion
  { (code) in
```

File: Sources/ValCommand/ValCommand.swift [hunk 10]
```diff
@@ -203,8 +203,18 @@ public struct ValCommand: ParsableCommand {
...
 
+    let codeFormatter: CodeTransform? = (try? find("clang-format")).map({
+      clangFormatter(URL(fileURLWithPath: $0))
+    })
+
+    //       core: program.cxx(program.corelib!, withFormatter: codeFormatter),
+    //       source: program.cxx(program[sourceModule], withFormatter: codeFormatter)
+    // =======
+    //       core: program.cxx(program.coreLibrary!), source: program.cxx(program[sourceModule])
+    // >>>>>>> upstream/main
```
Comment (kyouko-taiga) [inline Sources/ValCommand/ValCommand.swift:214]: What's this?

**Review by kyouko-taiga — COMMENTED** [review]
There seems to be leftover comments in the PR. Apart from that and a couple of minor suggestions, LGTM.

File: Sources/ValCommand/ValCommand.swift (hunk 10 above)
Comment (lucteo) [reply Sources/ValCommand/ValCommand.swift]: :( apparently I didn't properly save the file.

**Pull request merged**